src/services/actions-service/actions-service.ts

```diff
@@ -157,4 +157,29 @@ class ActionsService implements IActionsService {
     }
     return actions;
   }
+
+  sortActions(actions: Action[]): Action[] {
+    return actions.sort((a, b) => {
+      // Sort by sequence number
+      if (
+        a.transactionInfo.sequenceNumber !== b.transactionInfo.sequenceNumber
+      ) {
+        return (
+          a.transactionInfo.sequenceNumber - b.transactionInfo.sequenceNumber
+        );
+      }
+
+      // Sort by account update index within the transaction
+      const aAccountUpdateIndex =
+        a.transactionInfo.zkappAccountUpdateIds.indexOf(
+          Number(a.accountUpdateId)
+        );
+      const bAccountUpdateIndex =
+        b.transactionInfo.zkappAccountUpdateIds.indexOf(
+          Number(b.accountUpdateId)
+        );
+
+      return aAccountUpdateIndex - bAccountUpdateIndex;
+    });
```

Comment:
Shouldn't this include action ids and field element ids also? The protocol supports multiple actions per account update

Reply:
Also, are we not doing similar sorting for events?

Reply:
> Shouldn't this include action ids and field element ids also? The protocol supports multiple actions per account update

Within one account update, the actions are ordered within the `data` property in the response, see: https://github.com/o1-labs/Archive-Node-API/pull/107/files#diff-ce4b532132cbf96cc52ea84724fbfa50698c62ca9c690844068dafbfc687d3a8R341

```ts
const structData = eventData.data;
assert.strictEqual(structData.length, 16);
assert.strictEqual(structData[0], '2');
assert.deepStrictEqual(
  structData.slice(1, 6),
  structToAction(expectedS1)
);
assert.deepStrictEqual(
  structData.slice(6, 11),
  structToAction(expectedS2)
);
assert.deepStrictEqual(
  structData.slice(11, 16),
  structToAction(expectedS3)
);
```

Reply:
I haven't changed the intra-AU event order here, as it seems to be correct already.

Reply:
> Also, are we not doing similar sorting for events?

I've always been told as a zkapp developer that event order is not guaranteed, so I don't think we need to.  The only reason we need to sort actions is because an actions hash is committed to, and that hash uses a particular order that needs to be reproduced.  There is no such events hash on chain, so no ordering should be necessary.

Reply:
The query [here](https://github.com/o1-labs/Archive-Node-API/blob/9dcee361d812024798e40e02f409c9763e89c847/src/db/sql/events-actions/queries.ts#L160) to fetch actions uses `ANY` and so the ordering is not guaranteed. I'm guessing the ordering here just happens to be correct

Reply:
Ok I'll double check

Reply:
@deepthiskumar we convert field arrays into `data` [here](https://github.com/o1-labs/Archive-Node-API/blob/9dcee361d812024798e40e02f409c9763e89c847/src/services/data-adapters/database-row-adapters.ts#L204).  It seems like we take the `element_ids` for the correct order, and `elementIdFieldValues` (key-value map) to access the correct values.

Maybe I can add a test to make sure this order is explicitly respected, but I think it happens to work correctly already.

Reply:
Could you please add such a test? I think that'll help a lot with order being preserved implicitly.  Everything else looks good otherwise.

Reply:
Yes I will add a test in a fast-follow!